enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,390 @@
+	// name of the referent.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+	// +optional
+	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+}
+```
+
+_Note_: The secret is required to be created in the same namespace as that of the Route.
+The secret must be of type `kubernetes.io/tls` and the tls.key and the tls.crt key must be
+provided in the `data` (or `stringData`) field of the Secret configuration.
+
+If neither `.spec.tls.externalCertificate` or `.spec.tls.certificate` and `.spec.tls.key` are
+provided the router will serve the default generated certificates. User's will not be able to
+provide both `.spec.tls.certificate/key` and `.spec.tls.externalCertificate`. API server
+admission validation will enforce this.
+
+All valid and invalid scenarios will be depicted via the existing `RouteIngressCondition`.
+
+#### Variation
+
+N.A
+
+### Implementation Details/Notes/Constraints [optional]
+
+The router will read the secret referenced in `.spec.tls.externalCertificate` and will use
+the certificate inside to configure HAProxy if the secret is present and if the
+following pre-conditions are met:
+
+- Validations done by the router as part of [ExtendedValidateRoute()](https://github.com/openshift/router/blob/c407ebbc5d8d85daea2ef2d1ba539444a06f4d25/pkg/router/routeapihelpers/validation.go#L158) (contents of secret),
+
+  - The secret created should be in the same namespace as that of the route.
+  - The secret created is of type `kubernetes.io/tls`.
+  - Verify certificate and key (PEM encode/decode)
+  - Verify private key matches public certificate
+
+- Validations done by API server as part of [ValidateRoute()](https://github.com/openshift/openshift-apiserver/blob/aac3dd5bf0547e928103a0f718ca104b1bb13930/pkg/route/apis/route/validation/validation.go#L21),
+
+  - The router serviceaccount must have permission to read this secret particular secret.
+    - The role and rolebinding to provide this access must be provided by the user.
+  - CEL validations and o/library-go will enforce that both `.spec.tls.certificate` and `.spec.tls.externalCertificate`
+    are not specified on the route.
+
+The router being an edge component, from a security standpoint is more prone to
+being compromised. In order to avoid providing the router with escalated privileges
+to read all secrets, the router will implement a single item list/watch for secrets (secret monitor).
+This uses name-scoped rbac (created by the user) to access the particular secrets.
+
+A watch based secret monitor will be introduced in the router in order to keep
+track of all the secrets referenced by the routes. This component is solely
+responsible for maintaining all the single item list-watch functions required
+to cache the referenced secrets.
+
+The router will bootstrap the secret monitor to ensure it can keep the
+certificate/secret up-to-date. This means the router pod will maintain active watches
```

Comment:
```suggestion
certificate up-to-date. This means the router pod will maintain active watches
```
I suppose that the sectet will be kept up to date by the external certificate manager. The service monitor will solely read the secret and maintain the HAProxy certificate up to date.